[Peek] Wrong File/Folder Size

Deepak-Sangle opened the pull request:
## Summary of the Pull Request
When showing the file/folder size, Peek shows the value in power of 1024, but shows unit in power of 1000.
## PR Checklist

- [ ] **Closes:** #27871
- [x] **Communication:** I've discussed this with core contributors already (CC @PesBandi )
- [x] **Tests:** No need
- [x] **Localization:** In the file [src/modules/peek/Peek.UI/Strings/en-us/Resources.resw](https://github.com/microsoft/PowerToys/blob/a6f549bb5ad19d9382adc68c5da32e4f03b43fd3/src/modules/peek/Peek.UI/Strings/en-us/Resources.resw), the string `ReadableString_BytesString` and `ReadableString_ByteString` should not  be localized (cc @jaimecbernardo)
- [x] **Dev docs:** No need
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [x] **Documentation updated:** No need

## Detailed Description of the Pull Request / Additional Comments
I am displaying the file/folder size in a format similar to this- 10 KB (10,000 bytes).

## Validation Steps Performed
Simply observing files of different sizes after building the solution suffices for validation.

Deepak-Sangle committed:
[Peek] displaying file size in correct grammatical format

Deepak-Sangle committed:
Update Directory.Packages.props

Deepak-Sangle committed:
removed unnecessary file

Deepak-Sangle committed:
Merge branch 'fix/peek/file-size' of https://github.com/Deepak-Sangle/PowerToys into fix/peek/file-size

Deepak-Sangle committed:
[Peek] Added new symbols in resources.resw

Deepak-Sangle committed:
Merge branch 'main' of https://github.com/Deepak-Sangle/PowerToys into fix/peek/wrongSize

Deepak-Sangle committed:
[Peek] added commas separation

htcfreek commented:
I think if we do this we should ...
- switch the values and have KiB in brakets.
- should add an info tool tip why the values are different to the on in Explorer.

I worry that this confuses our users because the Explorer does it wrong since ever.

PesBandi commented:
I don't think you need the `totalBytesDisplays` variable. That should be always the same as `format[0]`, right?

Deepak-Sangle commented:
> switch the values and have KiB in brakets
I think, for folders of sizes in Gigabytes, the format won't be very user friendly.
> should add an info tool tip why the values are different to the on in Explorer
Well, the values are same as that of the explorer, only the unit is changed
> I don't think you need the totalBytesDisplays variable. That should be always the same as format[0], right?
Umm, format[0] = "{0} bytes", whereas totalBytesDisplays = "({1} bytes)", I can't find a way to use the same variables.

Deepak-Sangle committed:
modified spell check

PesBandi commented:
> Umm, format[0] = "{0} bytes", whereas totalBytesDisplays = "({1} bytes)", I can't find a way to use the same variables.
I'm sorry for my mistake, I somehow didn't realize those were two very different strings.

Also, why did you decide to use KiB? What (except for explorer) uses KiB? I think we should do something like "1 KB (1000 bytes)", since most other apps that I know use KB and it would be nice to be consistent with the rest of the world. And if you're concerned about explorer showing something else than Peek, that's what the exact number bytes part is for.

Deepak-Sangle committed:
[Peek] Changed 1024 factor to 1000 factor

Deepak-Sangle committed:
modified spell check

Deepak-Sangle commented:
Well, I get your point. I refactored the code to use 1000 instead of 1024. 
One thing to note is that I have used the "N0" separator, which basically is used to add commas in thousand's place of the total number of bytes section. Thus, I think it would always show a number like that even when someone is from a region (e.g. France) where the "dot" is the separator or from a region (like India) where the separation is done every hundred places.

PesBandi commented:
Isn't that because `string.Format` is used with `CultureInfo.InvariantCulture`? It should be formatted correctly if you don't use Invariant Culture (I am really not sure and I may be very wrong about this).

Deepak-Sangle commented:
Yes, I think `CultureInfo.InvariantCulture` is used to display consistent strings for dates, file size, etc. across different regions but I am not sure since I cannot test this and I don't know what Microsoft Powertoys wants me to do (make it consistent across regions or make it vary region wise).

htcfreek commented:
> Yes, I think `CultureInfo.InvariantCulture` is used to display consistent strings for dates, file size, etc. across different regions but I am not sure since I cannot test this and I don't know what Microsoft Powertoys wants me to do (make it consistent across regions or make it vary region wise). 

The correct thing would be to use `CultureInfo.CurrentCulture` as this is shown to the user.

htcfreek commented:
A screenshot of the current change would be helpful.

Deepak-Sangle commented:
Currently, Peek shows files similar to `10 MB (100,000,000 bytes)` as I used `CultureInfo.InvariantCulture`. I also tried using `CultureInfo.CurrentCulture`, and it showed me the correct format for India and I believe it will show country-specific strings correctly. 

Thus, now I want to ask what are the guidelines to follow. Should I display the string consistently following US format (100,00,000) or make it such that the string should be displayed country-specific?

Deepak-Sangle committed:
made string country specific

DeepakSangle commented:
I have done all the changes required. Can someone please merge this pull request? Thank you.

Deepak-Sangle commented:
I have completed all the changes that were required. Can someone please review and merge this pull request? Thank you.

jaimecbernardo commented:
Hi, Thanks for opening this PR.
As asked before, can we please get some screenshots to understand the differences? Thank you!

jaimecbernardo commented:
Also, please add to the resource comments that `{1:N0}` shouldn't be localized or it may cause application crashes.

Deepak-Sangle commented:
This is how the Peek shows the folder sizes. The commas are like that because in India, the given custom is followed. It will be region specific.
<img src="https://github.com/microsoft/PowerToys/assets/78836762/b50ff0fd-8514-465c-a23a-730cef98d7f6" textAlign="center" width="500"> </img>

> Also, please add to the resource comments that {1:N0} shouldn't be localized or it may cause application crashes.

I do not know where I have to add. Should I add it to the `Localization` point in my PR comment?

jaimecbernardo reviewed (commented):
Thank you for the contribution.
Given that the idea of Peek is being integrated into Explorer, I think it should still calculate values for the MB part as explorer does, dividing by 1024 instead of 1000. The strings in resources can be changed to KiB, MiB, etc. if you think that makes more sense, but the values shown should be similar to explorer.
@SamChaps , any comment here?

PesBandi commented:
>Given that the idea of Peek is being integrated into Explorer, I think it should still calculate values for the MB part as explorer does, dividing by 1024 instead of 1000. The strings in resources can be changed to KiB, MiB, etc. if you think that makes more sense, but the values shown should be similar to explorer.

Hello @jaimecbernardo, I understand your point about Peek being integrated into explorer and I also understand why you want the values shown to be similar to what explorer shows. I personally think that showing the exact number of bytes should be enough to clear up any confusion and I can't name a single piece of software (other than Windows utils that measure in KiB and show KB) that uses KiB as it's primary unit of measurement (I'm not an expert in this field, but from my own experience I think almost everyone uses KB).

Deepak-Sangle commented:
Well, I support using KiB instead of KB. 
The reason is that, generally speaking, whenever a user sees a file size as KB, they interpret it as KiloByte but they think its of order of 2^10 bytes and not 1000 bytes. I myself before starting this issue used to think the same and I am talking on behalf of general public. 
Also showing KiB has an advantage as you can show the user both the file size in power of 1024 as well as power of 1000. Lemme explain, if a user sees 174.04 MiB (182,495,287 bytes), they can easily interpret the file size as 182.495 MB, but if we MB, they cannot determine MiB just by seeing the total number of bytes (They have to divide the number by 1024's powers).

jaimecbernardo commented:
@PesBandi 
I've tried checking some common software I use.
GitHub seems to do 1024:
![image](https://github.com/microsoft/PowerToys/assets/26118718/d0b850d8-0eca-46cd-b63f-d87b7ca5a8f6)

Chrome as well:
![image](https://github.com/microsoft/PowerToys/assets/26118718/e0bb845a-cd45-4126-ab9e-d2ca6f5c608e)

Firefox as well:
![image](https://github.com/microsoft/PowerToys/assets/26118718/3b1b320b-fc14-4b3a-a79b-4da4c34d6383)

Visual Studio Code:
![image](https://github.com/microsoft/PowerToys/assets/26118718/298ae885-1e50-40d9-b607-7acba71b7abf)

I've checked some games in Steam, for the size that it says they're taking on disk and the installation size matches what's show in Windows.

Perhaps I lucked out with what I've tested, but given that this will integrate with Explorer, it should try to follow a similar logic.

What do you think?

PesBandi commented:
Wow, I somehow always thought Chrome and VS Code show it the right way, not the explorer way. Thanks for pointing that out.

> Also showing KiB has an advantage as you can show the user both the file size in power of 1024 as well as power of 1000. Lemme explain, if a user sees 174.04 MiB (182,495,287 bytes), they can easily interpret the file size as 182.495 MB

I absolutely agree, very good point.

Let's do "KiB (B)"

htcfreek commented:
>> Also showing KiB has an advantage as you can show the user both the file size in power of 1024 as well as power of 1000. Lemme explain, if a user sees 174.04 MiB (182,495,287 bytes), they can easily interpret the file size as 182.495 MB

> I absolutely agree, very good point.

> Let's do "KiB (B)"

Agree. 

We could add an information to our docs on learn.microsoft.com about the difference.

Deepak-Sangle commented:
Hey @jaimecbernardo, @PesBandi. I made the necessary changes as discussed. Following is a screenshot of how it looks like.

<img src="https://github.com/microsoft/PowerToys/assets/78836762/5290041e-3665-4b25-929a-e3be92f3175d" width="600"> </img>
I request you to please approve the PR. 
Thanks.

crutkas commented:
Why is this getting shifted to MiB versus MB?  Consistency with File Explorer and well known patterns I think is key.

crutkas commented on src/modules/peek/Peek.UI/Strings/en-us/Resources.resw:
```diff
@@ -157,28 +157,28 @@
   </data>
-  <data name="ReadableString_KiloByteAbbreviationFormat" xml:space="preserve">
-    <value>{0} KB</value>
+  <data name="ReadableString_KibiByteAbbreviationFormat" xml:space="preserve">
```
I do not agree with change.  Please revert

jaimecbernardo replied:
@Deepak-Sangle , let's please revert this change. Even thought it would be more correct, every software I provided as a sample that shows MiB numbers still presents them as "MB", so we should follow conventions as well. My bad for suggesting the change.

DeepakSangle replied:
Done with the changes, Peek now again show something like below image

<img src="https://user-images.githubusercontent.com/78836762/265678620-b50ff0fd-8514-465c-a23a-730cef98d7f6.png" width="600"></img>

crutkas replied:
It should be consistent with file explorer's calculated value and unit style.  In FE, it would be MB, not MiB.  It is what people expect

When I look at the image from the bug, one question that came to mind was this just a simple rounding issue as .76 would round up to .8.

![image](https://github.com/microsoft/PowerToys/assets/1462282/b14e3542-6fe5-47bb-9834-3fa8e726cecf)

I am ok with having bytes added like FE has

htcfreek replied:
@crutkas
It's a known bug of Explorer.exe that it shows wrong information: Calculation of MiB and shown as MB.

This bug exists since ever. But I think this is something that completely confuses the users if MS will fix it.

crutkas replied:
Then that should be a user option to override but as of now this should match FE in terms of values and units being displayed otherwise it will confuse users. Let’s fix this up and add in the setting then in a different PR to let the user pick

crutkas commented:
@jaimecbernardo i removed my blocking as the change was reverted.  Values should be consistent with FE is my critical feedback.

jaimecbernardo reviewed (changes requested):
Size shown is still different than what's shown in File Explorer. Can we please still use 1024 in the calculations?

Deepak-Sangle commented:
@jaimecbernardo @crutkas. If I understood the requirement correctly, the frontend should display the calculated values as well as the unit, the same as that of the explorer. That means calculating values using 1024 and showing units in KB, MB, etc. Correct?

> When I look at the image from the bug, one question that came to mind was this just a simple rounding issue as .76 would round up to .8.

For this issue, I found out that File Explorer shows numbers with different precision. The table below justifies it.
Total Bytes | 1023 | 1562 | 12,288  | 95,006 | 1,03,566 
--- | --- | --- | --- |--- | ---
File Explorer displays  | 1023 bytes | 1.52 KB | 12.0 KB | 92.7 KB | 101 KB 

It basically shows a maximum of 3 numbers for units of KB, MB, etc. So, it doesn't round for a number greater than or equal to 100 KB, rounds to 1 decimal for numbers between 10 to 99 KB, and rounds to 2 decimal places for numbers less than 10 KB.

So, should I maintain this system as well in Peek (This would require adding considerate amount of code) ?

Deepak-Sangle committed:
fix: used 1024 and fixed precision

Deepak-Sangle commented:
So, I have changed the code to use 1024 to calculate the values and display units in KB, MB, etc. I have also fixed the precision issue. Following are some of the boundary cases that I had to look into while solving it.

Total Bytes | What Peek and File Explorer <br>shows currently | Comments
--- |--- | ---
0 | 0 bytes | There are no brackets to display 0 bytes again
1 | 1 byte (1 byte) | This is the only case where Peek and File Explorer shows different values. <br> Reason is this PR: #27869 <br> Difference: Peek does not show the extra `s` at the end of the word bytes.
1023 | 1023 bytes (1,023 bytes) | None
3292 | 3.21 KB (3,292 bytes) | Displays number up to 2 decimal precision
16805| 16.4 KB (16,805 bytes) | Displays number up to 1 decimal precision
998869 | 975 KB (9,98,869 bytes) | Displays number up to 0 decimal precision
1025684 | 0.97 MB (10,25,684 bytes) | This used to display 1001.6 MB in the previous version of Peek <br> i.e., numbers in the range of 1000 to 1023 KB and after that
1061956 | 1.01 MB (10,61,956 bytes) | As expected. From now on, the same pattern repeats.

Also, note that FIle Explorer actually does not round off the decimal places but just takes the first n digits after the decimal point. For example, 1025684 bytes = 0.978 MB = 0.97 MB in File Explorer instead of 0.98 MB by rounding off.

Evidence for the same is attached here.

<details>
<summary>0 bytes</summary>
<br>
<img src="https://github.com/microsoft/PowerToys/assets/78836762/9d426365-5e6a-4852-810b-e66a61ec8a5a" ></img>
</details>

<details>
<summary>1 byte</summary>
<br>
<img src="https://github.com/microsoft/PowerToys/assets/78836762/7aecdd8f-3c22-4b81-9056-b51e23fe4940" ></img>
</details>

<details>
<summary>3.21 KB</summary>
<br>
<img src="https://github.com/microsoft/PowerToys/assets/78836762/4caed24c-8e00-4b38-a817-cea82a34eebb" ></img>
</details>

<details>
<summary>16.4 KB</summary>
<br>
<img src="https://github.com/microsoft/PowerToys/assets/78836762/b9f610ee-25d1-4c1d-b59f-a4c3587ccdff" ></img>
</details>

<details>
<summary>975 KB</summary>
<br>
<img src="https://github.com/microsoft/PowerToys/assets/78836762/c719ea33-918f-4ded-92f2-45a523decfc1" ></img>
</details>

<details>
<summary>0.97 MB</summary>
<br>
<img src="https://github.com/microsoft/PowerToys/assets/78836762/095f5cda-35f7-47bf-9c90-fcd23be6d247" ></img>
</details>

<details>
<summary>1.01 MB</summary>
<br>
<img src="https://github.com/microsoft/PowerToys/assets/78836762/f73a0aac-74d3-48fd-a935-a8943402c03d" ></img>
</details>

Deepak-Sangle committed:
spellcheck fixed

htcfreek commented:
@Deepak-Sangle 
Can you please provide updates screenshots that we can see how the precision thing si handled now.

DeepakSangle commented:
> @Deepak-Sangle Can you please provide updates screenshots that we can see how the precision thing si handled now.

@htcfreek I have attached the screenshots in my above comment. Please click on each dropdown item denoting the sizes (0 bytes, 1 bytes, etc.)

htcfreek reviewed (approved):
Thank you for the contribution. LGTM.

jaimecbernardo commented on src/modules/peek/Peek.UI/Strings/en-us/Resources.resw:
```diff
@@ -253,4 +253,12 @@
   </data>
+  <data name="ReadableString_BytesString" xml:space="preserve">
+    <value> ({1:N0} bytes)</value>
+    <comment>Displays total number of bytes</comment>
```
```suggestion
    <comment>Displays total number of bytes. Don't localize the "{1:N0}" part.</comment>
```

jaimecbernardo commented on src/modules/peek/Peek.UI/Strings/en-us/Resources.resw:
```diff
@@ -253,4 +253,12 @@
+  </data>
+  <data name="ReadableString_ByteString" xml:space="preserve">
+    <value> ({1:N0} byte)</value>
+    <comment>Displays unit byte</comment>
```
```suggestion
    <comment>Displays unit byte. Don't localize the "{1:N0}" part.</comment>
```

jaimecbernardo reviewed (approved):
LGTM! Thank you for the changes! I'm adding some comments to avoid localization mistakes, but other than that it looks great.
Thanks for the contribution!

jaimecbernardo committed:
Update src/modules/peek/Peek.UI/Strings/en-us/Resources.resw

jaimecbernardo committed:
Update src/modules/peek/Peek.UI/Strings/en-us/Resources.resw
